File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long n;
    //Prompt user for credit card #
    do
    {
        n = get_long("Enter credit card number.\n");
    }
    while (n < 0);

    //Finding every other digit from second last, x2, and add digits
    int f;
    for (long i = 100; i < n * 10; i = i * 100)
    {
        long a = n % i;
        int b = a / (i / 10);
        int c = b * 2;
        int d = c % 10;
        int e = c / 10;
        f = f + e + d;
    }

    //For adding up all other digits not multiplied and add together
    int j;
    for (long x = 10; x < n * 10; x = x * 100)
    {
        long g = n % x;
        int h = g / (x / 10);
        j = j + h;
    }

    //Find check sum
    int check = f + j;
    printf("Sum Check:%i\n", check);
    int k = check % 10;

    //Finding first 2 digits of credit card
    long m = n;
    long q = 2;
    while (m >= 100)
    {
        m = m / 10;
        q = q + 1;
    }

    //Checking the type of credit card, first with check sum
    if (k > 0)
    {
        printf("INVALID\n");
    }
    //Checking if American Express
    else if (q == 15)
    {
        if (m == 34 || m == 37)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    //Checking if Visa
    else if (q == 13)
    {
        if (m >= 40 && m < 50)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    //Checking if Visa or Mastercard
    else if (q == 16)
    {
        if (m >= 40 && m < 50)
        {
            printf("VISA\n");
        }

        else if (m == 51 || m == 52 || m == 53 || m == 54 || m == 55)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    //If check sum is valid but doesn't belong to the 3 groups
    else if (q < 13 || q > 16)
    {
        printf("INVALID\n");
    }
}

File: pset6/credit/credit.py
# Determine what brand a credit card is from

from cs50 import get_string
import sys

# Prompt user for card number as a string (to access any digit directly)
card = get_string("Credit Number: ")
card = card.strip()
cl = len(card)

total = 0
digits = []
# Iterate over every other number starting from second last digit 
for i in range(cl - 2, -1, -2):
    # Mutiply digit by 2 and add to list
    digits.append(int(card[i]) * 2)

# Add the digits of the mutiplied value together
for i in range(len(digits)):
    if digits[i] < 10:
        total += digits[i]
    elif digits[i] > 10:
        total += 1 + (digits[i] - 10)
    else:
        total += 1
    
# Iterate over the rest of the number
for i in range(cl - 1, -1, -2):
    # Add remaining numbers
    total += int(card[i])

# If total doesn't end in a 0 than INVALID card
if not total % 10 == 0:
    sys.exit("INVALID")
# Checking for AMEX
elif cl == 15:
    if int(card[0] + card[1]) == 37 or int(card[0] + card[1]) == 34:
        sys.exit("AMEX")
    else:
        sys.exit("INVALID")
# Checking for MasterCard or Visa
elif cl == 16:
    if int(card[0] + card[1]) < 56 and int(card[0] + card[1]) > 50:
        sys.exit("MASTERCARD")
    elif int(card[0]) == 4:
        sys.exit("VISA")
    else:
        sys.exit("INVALID")
# Checking for Visa
elif cl == 13:
    if int(card[0]) == 4:
        sys.exit("VISA")
    else:
        sys.exit("INVALID")
else:
    sys.exit("INVALID")